azavea/raster-foundry
# Tool Metadata + ToolRun Overrides

## Overview

This PR adds the `NodeMetadata` type, which will serve as an (optional) means of specifying preferences vis-a-vis rendering tiles and labelling/describing the semantics of AST nodes.

### Notes

I ended up running into some difficulties due to the serialization of ASTs and ToolRuns (essentially, they weren't round-tripping without loss and overriding values storied in the database could under certain circumstances lead to loss of data integrity). In response to this, I added some property based testing to ensure the validity of our `Encoder`/`Decoder` pairs.


## Testing Instructions

 * Create a Tool and ToolRun in the database. Here are some example POST bodies (note that the `ToolRun` refers to the `Tool` and that the `ToolRun`'s 'sources' block consists of mappings from IDs which specify nodes of the referent AST to a scene id and a band):
```json
{
	"organizationId": "dfac6307-b5ef-43f7-beda-b9f208bb7726",
	"title": "test",
	"description": "a simple, local-addition AST",
	"requirements": "",
	"license": "",
	"visibility": "public",
	"compatibleDataSources": [""],
	"stars": 5.5,
	"definition": {
    "id": "e76c3198-93a0-4a22-9d58-4ba645f8d6f4",
    "args": [
      {
        "id": "a9f4066b-a783-46b6-997d-a0e0d377a1ec"
      },
      {
        "id": "580dc0c8-9454-43f5-b72f-1854d7c18e13"
      }
    ],
    "apply": "+"
  },
	"tags": [],
	"categories": []
}
```
```json
{
  "visibility" : "PUBLIC",
  "organizationId" : "dfac6307-b5ef-43f7-beda-b9f208bb7726",
  "tool" : <TOOL-ID>,
  "executionParameters" : {
        "sources": {
          "a9f4066b-a783-46b6-997d-a0e0d377a1ec": {
            "id": <SOME_SCENE_ID>,
            "band": 4,
            "type": "scene"
          },
          "580dc0c8-9454-43f5-b72f-1854d7c18e13": {
            "id": <SOME_SCENE_ID>,
            "band": 5,
            "type": "scene"
          }
        }
      }
}
```
 * Assuming you've provided meaningful (and ingested) scene information, you should be able to see a tile at `http://localhost:9900/tiles/tools/<TOOLRUN_ID/{z}/{x}/{y}/?token=<RF_TOKEN>`
 * To test out per-node viewing, simply add the query parameter node + the UUID of the node you'd like to see TMS rendered (`&node=<NODE_ID>`)

Closes #1561
Closes #1555
